### ISSUE TITLE
Index 0 is out of range error from `TAuthorizationRuleCollection::insertAt`

### ISSUE THREAD

**Issue by ganiuszka** [pradosoft/prado#810]
Hello The PRADO Group,

I have a problem with my application after upgrade to PRADO 4.2.0. The application uses own config to store user authorization properties (roles, ips). A web user manager applies the authorization rules on the `onAuthorization` event. In short it looks like below:

```
$auth_rules = $this->getApplication()->getAuthorizationRules();

// remove authorization rules if any
$auth_rules->clear();

[ ...HERE IS CODE THAT PREPARES RULES... ]

// Add authorization rules
for ($i = 0; $i < count($rules); $i++) {
        $rule = new TAuthorizationRule(
		$rules[$i]['action'],
		$rules[$i]['users'],
		$rules[$i]['roles'],
		$rules[$i]['verb'],
		$rules[$i]['ips']
        );
        $auth_rules->insertAt($i, $rule);
}
```

On PRADO 4.2.0 the authorization finishes with the following error:

```
Prado\Exceptions\TInvalidDataValueException
Description
Index '0' is out of range.
```

I looked a bit in the code and I found that the `insertAt()` method in `TAuthorizationRuleCollection` class stopped working:

```
public function insertAt($index, $item)
{
        if ($item instanceof TAuthorizationRule) {
                parent::insertAt($index, $item);
         } else {
                throw new TInvalidDataTypeException('authorizationrulecollection_authorizationrule_required');
         }
}
```

It is because in previous versions the `TAuthorizationRuleCollection` extended `TList`:

```
class TAuthorizationRuleCollection extends \Prado\Collections\TList
```

and now it exteds `TPriorityList`:

```
class TAuthorizationRuleCollection extends \Prado\Collections\TPriorityList
```

The problem starts in `TPriorityList` in the `insertAt()` method which expects that the priority will exists in the list before adding a new item. The TList does not have this requirement and from this reason ``TList::insertAt()`` worked correctly with `TAuthorizationRuleCollection`.

```
public function insertAt($index, $item)
{
        if ($this->getReadOnly()) {
                throw new TInvalidOperationException('list_readonly', get_class($this));
        }

        if (($priority = $this->priorityAt($index, true)) !== false) {
                $this->insertAtIndexInPriority($item, $priority[1], $priority[0]);
        } else {
                throw new TInvalidDataValueException('list_index_invalid', $index);
        }
}
```

Would it be possible to fix the `TAuthorizationRuleCollection::insertAt()` method?

Thank you in avance for your help.

Best regards,
Marcin Haba (gani)

**Comment by ganiuszka** [comment]
In my previous message I provided not completed error message. Here is the full message:
```
Prado\Exceptions\TInvalidDataValueException
Description
Index '0' is out of range.

Source File
${PradoFramework}/Collections/TPriorityList.php (550)
```

**Comment by ctrlaltca** [comment]
I've committed a quick fix in aee2a29b596b2e4d25310a9f88bdb2fbdae6db90
Can you please confirm if it fixes the problem for you?
Thank you

**Comment by ganiuszka** [comment]
Hello ctrlaltca,

Thank you for the quick fix.

Yes I confirm that it fixes this problem.

Many thanks.

Best regards,
Marcin Haba (gani)

**Comment by ctrlaltca** [comment]
Just released version 4.2.1 with the fixes.
